src/main/java/org/primefaces/component/datatable/DataTable.java

```diff
@@ -665,31 +668,29 @@ public String resolveStaticField(ValueExpression expression) {
         return null;
     }
 
-    public String resolveDynamicField(ValueExpression expression) {
-
-        if (expression == null) {
+    /**
+     * Get bean's property value from a value expression.
+     * Support old syntax (e.g #{car[column.property]}) instead of #{column.property}
+     * @param exprVE
+     * @return
+     */
+    public String resolveDynamicField(ValueExpression exprVE) {
+        if (exprVE == null) {
             return null;
         }
 
         FacesContext context = getFacesContext();
         ELContext elContext = context.getELContext();
+        String exprStr = exprVE.getExpressionString();
 
-        String expressionString = expression.getExpressionString();
-
-        // old syntax compatibility
-        // #{car[column.property]}
-        // new syntax is:
-        // #{column.property} or even a method call
-        if (expressionString.startsWith("#{" + getVar() + "[")) {
-            expressionString = expressionString.substring(expressionString.indexOf('[') + 1, expressionString.indexOf(']'));
-            expressionString = "#{" + expressionString + "}";
-
-            ValueExpression dynaVE = context.getApplication()
-                    .getExpressionFactory().createValueExpression(elContext, expressionString, String.class);
-            return (String) dynaVE.getValue(elContext);
+        Matcher matcher = DYNAMIC_FIELD_REGEX.matcher(exprStr );
+        if (matcher.find()) {
+            exprStr = matcher.group(1);
+            exprVE = context.getApplication().getExpressionFactory()
+                    .createValueExpression(elContext, "#{" + exprStr  + "}", String.class);
         }
```

Comment:
Maybe this whole block can be a sub method and you can write unit tests that show all the permutations of of the REGEX being parsed.  Or maybe just break the regex parse and matcher.group(1) return into a method so we can unit test just that?

Reply:
That's something you can do already in `DataTableTest#testResolveDynamicField`